File: Java/src/deepspace/SpaceCity.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

public class SpaceCity extends SpaceStation {
    SpaceStation base;
    ArrayList<SpaceStation> collaborators;
    
    SpaceCity(SpaceStation b, ArrayList<SpaceStation> rest) {
        super(b.getName(), new SuppliesPackage(b.getAmmoPower(), b.getFuelUnits(), b.getShieldPower()),
            b.getNMedals(), b.getPendingDamage(), b.getWeapons(), b.getShieldBoosters(), b.getHangar());
        base = b;
        collaborators = rest;
    }
    
    public ArrayList<SpaceStation> getCollaborators() {
        return collaborators;
    }
    
    public SpaceCityToUI getUIversion() {
        return new SpaceCityToUI(this);
    }
    
    @Override
    public String toString() {
        return getUIversion().toString();
    }
    
    @Override
    public float fire() {
        int suma = 0;
        for (SpaceStation c: collaborators) {
            suma += c.fire();
        }
        return suma + super.fire();
    }
    
    @Override
    public float protection() {
        int suma = 0;
        for (SpaceStation c: collaborators) {
            suma += c.protection();
        }
        return suma + super.protection();
    }
    
    @Override
    public Transformation setLoot(Loot loot) {
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }
}

File: Ruby/SpaceCity.rb
#encoding: UTF-8

require_relative 'SpaceStation'
require_relative 'SuppliesPackage'
require_relative 'Transformation'
require_relative 'SpaceCityToUI'

module Deepspace
	class SpaceCity < SpaceStation
		
		def initialize (base, rest) # SpaceCity(base : SpaceStation, rest : SpaceStation[])
			super(base.name, SuppliesPackage.new(base.ammoPower, base.fuelUnits, base.shieldPower), \
				base.nMedals, base.pendingDamage, base.weapons, base.shieldBoosters, base.hangar)
			@base = base
			@collaborators = rest
		end
		
		attr_reader :collaborators
		
		def getUIversion
        	SpaceCityToUI.new(self)
        end
        
        def to_s
        	getUIversion.to_s
      	end
		
		def fire
			suma = 0
			for c in @collaborators
				suma += c.fire
			end
			suma + super
		end

		def protection
			suma = 0
			for c in @collaborators
				suma += c.protection
			end
			suma + super
		end
		
		def setLoot(loot)
			super
			Transformation::NOTRANSFORM
		end
	end
end
